File: lab2_9_12/1.py
import sys

def dfs(row, col, graph, vis, dr, dc, path, n, m, directions):
    if row == n-1 and col == m-1:
        return True
    
    vis[row][col] = 1

    for i in range(4):
        nr, nc = row + dr[i], col + dc[i]

        if 0 <= nr < n and 0 <= nc < m and graph[nr][nc] == 1 and not vis[nr][nc]:
            path.append(directions[i])

            if dfs(nr, nc, graph, vis, dr, dc, path, n, m, directions):
                return True

            path.pop()

    return False

def main():
    try:
        line = input("Enter no. of rows & cols : ").split()
        if not line: return
        n, m = int(line[0]), int(line[1])

        print("Enter the grid : ")
        grid = []
        for _ in range(n):
            grid.append(list(map(int, input().split())))
            
        dr = [1, -1, 0, 0]
        dc = [0, 0, 1, -1]
        directions = "DURL"

        vis = [[0 for _ in range(m)] for _ in range(n)]
        path = []

        print("\nGrid entered is :")
        for row in grid:
            print(*(row))

        if grid[0][0] == 1:
            if dfs(0, 0, grid, vis, dr, dc, path, n, m, directions):
                print(f"Path found : {''.join(path)}")
            else:    
                print("Path not found")
        else:    
            print("Starting point blocked")
    except (ValueError, EOFError, IndexError):
        pass

if __name__ == "__main__":
    main()

File: lab2_9_12/1.cpp
#include<bits/stdc++.h>

using namespace std;

string dir = "DURL";

bool dfs(int row, int col, vector<vector<int>>& graph, vector<vector<int>>& vis, vector<int>& dr, vector<int>& dc, string& path, int n, int m)
{
    if(row == n-1 && col == m-1)
        return true;
    
    vis[row][col] = 1;


    for(int i=0;i<4;i++)
    {
        int nr = row + dr[i];
        int nc = col + dc[i];

        if(nr >= 0 && nr < n && nc >= 0 && nc < m && graph[nr][nc] == 1 && !vis[nr][nc])
        {
            path.push_back(dir[i]);

            if(dfs(nr, nc, graph, vis, dr, dc, path, n, m))
                return true;

            path.pop_back();
        }
    }

    return false;
}

int main()
{
    cout << "Enter no. of rows & cols : ";

    int n, m;
    cin >> n >> m;

    vector<vector<int>> grid(n, vector<int>(m));
    cout << "Enter the grid : \n";

    for(int i=0;i<n;i++)
        for(int j =0;j<m;j++)
            cin >> grid[i][j];
            
    vector<int> dr = {1, -1, 0, 0};
    vector<int> dc = {0, 0, 1, -1};

    vector<vector<int>> vis(n, vector<int>(m, 0));

    string path = "";

    cout << "\nGrid entered is :\n";
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<m;j++)
            cout << grid[i][j] << " ";
        cout << "\n";
    }

    if(grid[0][0] == 1)
    {
        if(dfs(0, 0, grid, vis, dr, dc, path, n, m))
            cout << "Path found : " << path << "\n";
        else    
            cout << "Path not found\n";
    }
    else    
        cout << "Starting point blocked\n";
    
        return 0;
}
